doc/barclamp.md

```diff
@@ -0,0 +1,1321 @@
+# fork, though. It also assumes that you are not working on your fork's
+# `master` branch (you shouldn't, but people have been known to do this
+# anyway). Since all of these may occasionally happen, this guide leaves it up
+# to the user to set $COMMIT as they see fit rather than trying to determine
+# its value automatically through the command mentioned above.
+
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
+
+# Insert your own barclamp's name here:
+export BARCLAMP_NAME=mybarclamp
+
+# Insert a space delimited list of the Barclamps you are working on here (only
+# required if your barclamp requires changes to other barclamps as well).
+
+export BARCLAMPS="$BARCLAMP_NAME"
+
+# Insert the path to your barclamp collection checkout here (in case it's
+# somewhere other than /root/crowbar-openstack). This directory must be a git
+# working copy of the crowbar-openstack fork and branch you are working on.
+
+export BARCLAMP_FORK_PATH=/root/crowbar-openstack
+
+export GIT_AUTHOR_NAME="$C_NAME"
+export GIT_AUTHOR_EMAIL="$C_EMAIL"
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="$C_EMAIL"
+
+# Applies all the committed changes in your crowbar-openstack working copy to
+# /opt/dell and updates Crowbar runtime state. Note: this will break if you
+# rebase your working copy, so do not rebase your working copy. If you rebase
+# it (e.g. for final testing before a merge, please build a fresh Crowbar node
+# or restore it from a snapshot).
+function reposync_crowbar()
+  {
+  if [ -e /opt/dell/patchlevel ]; then
+    . /opt/dell/patchlevel
+    head=$(cd $BARCLAMP_FORK_PATH; git log --format='%H'| head -n 1)
+    if [ "$head" = "$commit" ]; then
+      echo "/opt/dell is already up to date. If this is not what you " 1>&2
+      echo "expected to see, you may have uncommitted changes" 1>&2
+      echo "in ${BARCLAMP_FORK_PATH}." 1>&2
+      return
+    fi
+  else
+    commit="${COMMIT}^"
+  fi
+
+  pushd $BARCLAMP_FORK_PATH > /dev/null
+  cp -a /opt/dell /opt/dell.orig
+  git diff $commit HEAD | patch -N --merge -p1 -d /opt/dell || return
+  rm -rf /opt/dell.orig
+  # Record current HEAD as patchlevel
+  echo "commit=$(git log --format='%H'| head -n 1)" > /opt/dell/patchlevel
+  popd > /dev/null
+
+  for barclamp in $BARCLAMPS
+    do
+    if [ ! -h /opt/dell/crowbar_framework/barclamps/${barclamp}.yml ]; then
+      ln -s /opt/dell/${barclamp}.yml /opt/dell/crowbar_framework/barclamps 2>&1 > /dev/null
+    fi
+    done
+  chmod 755 /opt/dell/bin/*
+
+  sync_crowbar
+  }
+
+# Helper function: updates Crowbar's runtime state after /opt/dell
+# has been modified.
+sync_crowbar()
+  {
+  barclamp_install.rb /opt/dell/crowbar_framework/barclamps &&
+  for barclamp in $BARCLAMPS
+    do
+    knife cookbook upload -o /opt/dell/chef/cookbooks/ $barclamp
+    done
+  systemctl restart crowbar
+  }
+
+```
+
+We will make heavy use of the `reposync_crowbar` command thus defined in the
```

Comment:
"lowest common denominator" captures it pretty well, yes :-)

For any 10 developers you'll have about 25 vastly differing local development environments. Instead of trying to come up with something all of them are happy with I wanted to accomplish two things:

* Provide something anybody can use even if their workstation accidently got an anvil dropped on it an hour ago and now they have a new machine fresh from IT with none of their customized personal environment on it.
* Provide something that shows them all the bits they need to touch in a compact, easy to understand manner so they can roll their own development environment if they need to or fix the Guardfile themselves if it bitrots.